src/codegen.rs

```diff
@@ -46,81 +47,71 @@ impl<'a, 'b> TypeAndPointer<'a, 'b> {
     }
 
     /// returns the DataTypeInformation for the pointer's dereferenced type
-    pub fn get_type_information(&self) -> &DataTypeInformation<'a> {
-        self.type_entry.get_type_information().unwrap() //TODO
+    pub fn get_type_information(&self) -> &DataTypeInformation {
+        self.type_entry.get_type_information()
     }
 }
 
 /// the codegen struct carries all dependencies required to generate
 /// the IR code for a compilation unit
-pub struct CodeGen<'ctx> {
+pub struct CodeGen<'ink> {
     /// the LLVM context used to access the llvm typesystem, and create BasicBlocks
-    pub context: &'ctx Context,
+    pub context: &'ink Context,
     /// the module represents a llvm compilation unit
-    pub module: Module<'ctx>,
-    /// the builder is a reusable object to create ir-statements at a given location
-    pub builder: Builder<'ctx>,
+    pub module: Module<'ink>,
     /// the index / symbol table
-    pub index: &'ctx mut Index<'ctx>,
+    pub index: Index,
 }
 
-impl<'ctx> CodeGen<'ctx> {
+impl<'ink> CodeGen<'ink> {
+
     /// constructs a new code-generator that generates CompilationUnits into a module with the given module_name
     pub fn new(
-        context: &'ctx Context,
-        index: &'ctx mut Index<'ctx>,
+        context : &'ink Context,
+        index: Index,
```

Comment:
I think this is still a big problem.
I dont really see why it would need to own the index :-(